File: api/turing/models/router.go
package models

import (
	"database/sql"
	"encoding/json"
	"fmt"
)

type RouterStatus string

const (
	RouterStatusPending    RouterStatus = "pending"
	RouterStatusFailed     RouterStatus = "failed"
	RouterStatusDeployed   RouterStatus = "deployed"
	RouterStatusUndeployed RouterStatus = "undeployed"
)

// Router holds the information as to which versions of Router have been deployed,
// and acts as the top level object for an instance of the Turing Router.
type Router struct {
	Model
	// Project id of the project this router belongs to, as retrieved from
	// the MLP API.
	ProjectID ID `json:"project_id"`
	// Environment name of the environment this router belongs to, as retrieved
	// from the MLP API.
	EnvironmentName string `json:"environment_name"`
	// Name of the router. Must be unique within the given project and environment.
	Name string `json:"name"`
	// Status of the Router. Indicates the deployment status of the router.
	Status RouterStatus `json:"status"`
	// Endpoint URL where the currently deployed router version is accessible at
	Endpoint string `json:"endpoint,omitempty"`

	// The current version (may be deployed or undeployed)
	CurrRouterVersionID sql.NullInt32  `json:"-"`
	CurrRouterVersion   *RouterVersion `json:"config,omitempty" gorm:"foreignkey:CurrRouterVersionID"`

	// MonitoringURL is for all router versions
	MonitoringURL string `json:"monitoring_url" gorm:"-"`
}

func (r *Router) SetCurrRouterVersionID(routerVersionID ID) {
	r.CurrRouterVersionID = sql.NullInt32{Int32: int32(routerVersionID), Valid: true}
}

func (r *Router) ClearCurrRouterVersionID() {
	r.CurrRouterVersionID = sql.NullInt32{Int32: 0, Valid: false}
}

// SetCurrRouterVersion sets the currently version for this router to the provided
// routerVersion.
func (r *Router) SetCurrRouterVersion(routerVersion *RouterVersion) {
	r.SetCurrRouterVersionID(routerVersion.ID)
	r.CurrRouterVersion = routerVersion
}

// ClearCurrRouterVersion clears the current version for this router.
func (r *Router) ClearCurrRouterVersion() {
	r.ClearCurrRouterVersionID()
	r.CurrRouterVersion = nil
}

// RouterResponse is an alias for the Router, to enable custom marshaling
type RouterResponse Router

// MarshalJSON is a custom marshaling function for the Router, to manipulate the
// endpoint.
func (r *Router) MarshalJSON() ([]byte, error) {
	response := RouterResponse(*r)
	// Append the path to the endpoint
	if response.Endpoint != "" {
		response.Endpoint = fmt.Sprintf("%s/v1/predict", response.Endpoint)
	}
	return json.Marshal(response)
}

File: sdk/turing/router.py
from enum import Enum
from typing import Optional, List

import turing.generated.models
from turing._base_types import ApiObject, ApiObjectSpec


class RouterStatus(Enum):
    """
    Status of router
    """
    DEPLOYED = "deployed"
    UNDEPLOYED = "undeployed"
    FAILED = "failed"
    PENDING = "pending"


@ApiObjectSpec(turing.generated.models.Router)
class Router(ApiObject):
    """
    API entity for Router
    """
    def __init__(self,
                 id: int,
                 name: str,
                 project_id: int,
                 environment_name: str,
                 monitoring_url: str,
                 status: RouterStatus,
                 endpoint: str = None,
                 **kwargs):
        super(Router, self).__init__(**kwargs)
        self._id = id
        self._name = name
        self._project_id = project_id
        self._environment_name = environment_name
        self._endpoint = endpoint
        self._monitoring_url = monitoring_url
        self._status = RouterStatus(status)

    @property
    def id(self) -> int:
        return self._id

    @property
    def name(self) -> str:
        return self._name

    @property
    def project_id(self) -> int:
        return self._project_id

    @property
    def environment_name(self) -> str:
        return self._environment_name

    @property
    def endpoint(self) -> str:
        return self._endpoint

    @property
    def monitoring_url(self) -> str:
        return self._monitoring_url

    @property
    def status(self) -> RouterStatus:
        return self._status

    @classmethod
    def list(cls) -> List['Router']:
        """
        List routers in the active project

        :return: list of routers
        """
        response = turing.active_session.list_routers()
        return [Router.from_open_api(item) for item in response]
